Objects/weakrefobject.c

```diff
@@ -1017,10 +1102,29 @@ _PyStaticType_ClearWeakRefs(PyInterpreterState *interp, PyTypeObject *type)
 {
     static_builtin_state *state = _PyStaticType_GetState(interp, type);
     PyObject **list = _PyStaticType_GET_WEAKREFS_LISTPTR(state);
-    while (*list != NULL) {
-        /* Note that clear_weakref() pops the first ref off the type's
-           weaklist before clearing its wr_object and wr_callback.
-           That is how we're able to loop over the list. */
-        clear_weakref((PyWeakReference *)*list);
+    for (int done = 0; !done;) {
+        PyObject *callback = NULL;
+        LOCK_WEAKREFS(type);
+        if (*list != NULL) {
+            clear_weakref_lock_held((PyWeakReference*)*list, &callback);
+        }
+        done = (*list == NULL);
+        UNLOCK_WEAKREFS(type);
+        Py_XDECREF(callback);
+    }
```

Comment:
This is called at shutdown when there is only one thread left. We don't need to do any locking. I think the following code should be sufficient:

```
    PyObject **list = _PyStaticType_GET_WEAKREFS_LISTPTR(state);
    while (*list) {
        _PyWeakref_ClearRef((PyWeakReference *)*list);
    }
```

Reply:
This slightly changes the semantics - the original version decrefs the callback. Doing it this way postpones that decref until the weakref's destructor is invoked, which is probably fine.